trin-history/src/jsonrpc.rs

```diff
@@ -73,54 +79,132 @@ impl HistoryRequestHandler {
                     };
                     let _ = request.resp.send(response);
                 }
+                // todo: combine with TraceRecursiveFindContent & refactor once `break` from
+                // labeled blocks becomes stable
+                // https://blog.rust-lang.org/2022/11/03/Rust-1.65.0.html#break-from-labeled-blocks
```

Comment:
I spent quite some time (in vain) trying to combine & refactor these two rpc call execution blocks since they share a majority of code. However, it turns out that the early `return` pattern we used for parsing the params breaks the tx channel. I confirmed this by passing in invalid `RecursiveFindContent` params. Using an early `return` was about the only way I could think of to refactor this code. The other being the pattern linked above, but that's not a stable feature. If there are any other ideas on how to refactor this monstrosity of a nested code block, I'm all ears

Reply:
I think that `continue` can do what we need here.

Here's an approach that I tested (with some awkward safe unwraps that should be cleaned up) that combines both endpoints and minimizes nesting. 
```suggestion
...
  HistoryEndpoint::RecursiveFindContent
  | HistoryEndpoint::TraceRecursiveFindContent => {
      let params = RecursiveFindContentParams::try_from(request.params);
      if let Err(msg) = params {
          let _ = request.resp.send(Err(format!(
              "Invalid RecursiveFindContent params: {:?}",
              msg.code
          )));
          continue;
      }
      let params = params.unwrap();
      let content_key = HistoryContentKey::try_from(params.content_key.to_vec());
      if let Err(msg) = content_key {
          let _ = request
              .resp
              .send(Err(format!("Invalid content key requested: {}", msg)));
          continue;
      }
      let content_key = content_key.unwrap();
      // Check whether we have the data locally.
      let local_content: Option<Vec<u8>> =
          match self.network.overlay.store.read().get(&content_key) {
              Ok(Some(data)) => Some(data),
              Ok(None) => None,
              Err(err) => {
                  error!(
                      error = %err,
                      content.key = %content_key,
                      "Error checking data store for content",
                  );
                  None
              }
          };

      if local_content.is_some() {
          let response = match request.endpoint {
              HistoryEndpoint::RecursiveFindContent => {
                  let content = local_content.unwrap();
                  Value::String(hex_encode(content))
              }
              HistoryEndpoint::TraceRecursiveFindContent => {
                  json!({
                      "content": Value::Null,
                      "route": []
                  })
              }
              _ => json!("none"),
          };
          let _ = request.resp.send(Ok(response));
          continue;
      }

      let (content, closest_nodes) = self
          .network
          .overlay
          .lookup_content(content_key.clone())
          .await;

      let value = if request.endpoint == HistoryEndpoint::TraceRecursiveFindContent {
          let closest_nodes = closest_nodes.unwrap_or_default();
          let closest_nodes: Vec<Value> = closest_nodes
              .iter()
              // Skip over enrs that are unable to be looked up
              .filter_map(|node_id| {
                  let content_id = content_key.content_id();
                  // Return None for enrs that cannot be looked up
                  match self.network.overlay.discovery.find_enr(node_id) {
                      Some(enr) => {
                          XorMetric::distance(&content_id, &enr.node_id().raw())
                              .log2()
                              .map(|distance| {
                                  json!({
                                      "enr": enr,
                                      "distance": distance
                                  })
                              })
                      }
                      None => None,
                  }
              })
              .collect();

          json!({
              "content": match content {
                  Some(content) => Value::String(hex_encode(content)),
                  None => Value::Null,

              },
              "route": closest_nodes,
          })
      } else {
          match content {
              Some(content) => Value::String(hex_encode(content)),
              None => Value::Null,
          }
      };

      let _ = request.resp.send(Ok(value));
}
```